### DISCUSSION TITLE
Backing up Zimbra using restic

### DISCUSSION THREAD

**Discussion by ianw1974** [ianw1974/zimbra-build-scripts#69] [General]
# Backing up Zimbra using restic.

I've recently come across a cool backup program called restic, which can be used for various different scenarios.  I decided to see how useful this could be for backing up a Zimbra installation.

As most people will know, some files in Zimbra are sparse files - the LDAP database for example.  Therefore, this can be somewhat problematic if it isn't backed up and restored correctly.  I know that restic supports sparse files in it's newer versions, so it's best to be using the version downloaded from their Github releases, rather than use what the Linux distribution provides in it's repositories.  I downloaded this and extracted it to /usr/local/bin and renamed it to "restic".  Then set the execute bit ```chmod +x /usr/local/bin/restic``` and then we are ready to use it.

My testing is going to involve making backups with restic while Zimbra is running and also when it's been shut down.  This is to check if an online backup of Zimbra will actually be able to backup the LDAP sparse files and will it fail to start once it's been restored.

## Preparation

First, before we can start we need choose a place where to save the restic backups.  I have chosen ```/var/restic``` in my test here.  Once we have this location, we then need to initialise restic before we can start taking backups.  So:

```
root@zimbra:~/scripts# mkdir /var/restic
root@zimbra:~/scripts# restic init -r /var/restic
```

You will be prompted for a password when initialising restic.  For scripting purposes, we then need to store two values in a file.  The first is the directory where we initialised restic, the second being the password.  I usually create a hidden file called ```.restic-parms```, with values as below:

```
export RESTIC_REPOSITORY="/var/restic"
export RESTIC_PASSWORD="my-secret-password"
```

Obviously change the password to something more secure that you can easily remember.  For our backup script, we then add something like below:

```restic-backup-zimbra-online.sh ```
```
#!/bin/bash
#
# Backup Zimbra with restic
# Online backup

# Variables
BACKUP_PARMS=~/scripts/.restic-parms
BACKUP_DIR=/opt/zimbra

# Backup /opt/zimbra directory
echo "Backing up Zimbra..."
source ${BACKUP_PARMS}
restic backup ${BACKUP_DIR}
```

You can see that there are two variables - one for the restic parameters to load for the backup to be created, and the other for the directory that we wish to backup - in this case ```/opt/zimbra```.  As this is my online backup script, I am not stopping Zimbra before the backup.  The only difference between the two, is the end part of the script will look like this:

```
# Backup /home/ian directory
echo "Backing up Zimbra..."
source ${BACKUP_PARMS}
systemctl stop zimbra
restic backup ${BACKUP_DIR}
systemctl start zimbra
```

The online backup test is being done without any offline backups being taken first.  This will ensure that when we run the restore, it isn't going to attempt to pull any files from an offline snapshot.  This will be far more a valid test for ensuring the LDAP database is restored correctly.  Since if we did the initial backup offline, and then lots of online backups - we won't be sure if the LDAP DB came from an old backup or from the newer one.  This will confirm for 100% if the backup is actually sane.  Obviously if you can afford some downtime when making the backup, then you can stop Zimbra before running the restic backup process.  This would pretty much guarantee a successful backup in comparison to the online one, but we shall see how my test results perform.

## Initial backup

```
root@zimbra:~/scripts# ./restic-backup-zimbra-online.sh 
Backing up Zimbra...
repository 5339cb72 opened (version 2, compression level auto)
no parent snapshot found, will read all files

Files:       16442 new,     0 changed,     0 unmodified
Dirs:         1530 new,     0 changed,     0 unmodified
Added to the repository: 1.601 GiB (858.420 MiB stored)

processed 16442 files, 82.887 GiB in 2:30
snapshot b19544cf saved
```

Basically this is a clean install of Zimbra so no emails as such.  The script stops Zimbra, then creates the restic snapshot and starts Zimbra again.  As you can see above, the backup took 2mins 30 secs.  Let's take a look at disk space usage in comparison with the ```/opt/zimbra``` directory:

```
root@zimbra:~/scripts# du -sh /opt/zimbra
2.0G	/opt/zimbra

root@zimbra:~/scripts# du -sh /var/restic
860M	/var/restic
```

Restic will compress files, and it also uses deduplication.  This will become more apparent when we make subsequent backups.  When following backups are made, then it will be quicker than the initial one - especially when the system has a lot more emails than a clean install.  The reason is restic will compare the previous backup with the one being created and backup the changes.  To do a simulation by getting Zimbra to generate emails when it is stopped/started, this will at least give me some new files to backup without me having to manually send/receive emails.  This can be seen in the results below when we run another backup:

```
root@zimbra:~/scripts# ./restic-backup-zimbra-online.sh 
Backing up Zimbra...
repository 5339cb72 opened (version 2, compression level auto)
using parent snapshot b19544cf
[0:00] 100.00%  2 / 2 index files loaded

Files:         121 new,   223 changed, 16177 unmodified
Dirs:           18 new,   105 changed,  1405 unmodified
Added to the repository: 152.143 MiB (2.723 MiB stored)

processed 16521 files, 82.888 GiB in 2:08
snapshot 0af10032 saved
```

As you can see from the results, the backup has only taken a small amount of changed files within this backup - the emails we generated.  Unfortunately we have to ignore the way the Zimbra LDAP database is shown at 80GB because this is actually the sparse file which isn't actually using this amount of disk space.  We just have to accept that situation.  What is more important is shown as what is added to the repository - in this instance 152MB.  Here we can see the list of the backups (or snapshots as restic names them) below:

```
root@zimbra:~/scripts# restic snapshots
repository 5339cb72 opened (version 2, compression level auto)
ID        Time                 Host        Tags        Paths
------------------------------------------------------------------
b19544cf  2024-01-17 10:21:15  zimbra                  /opt/zimbra
0af10032  2024-01-17 10:33:54  zimbra                  /opt/zimbra
------------------------------------------------------------------
2 snapshots
```

Let's check the disk space used by restic after the second snapshot:

```
root@zimbra:~/scripts# du -sh /var/restic
863M	/var/restic
```

The size of ```/var/restic``` after the first backup was 860MB, the size after the second snapshot is 863MB.  So only 3MB additional backup space.  This is achieved by compression as well as deduplication.  In comparison, the /opt/zimbra directory takes 2GB.

## Removing backups

There may be a time where we decide we don't want to keep too many backups within restic, and there are two commands we can use for this.  First we would need to list the backups to obtain the ID of the snapshot that we want to delete.  So from the previous section we have two backups - let's say we want to delete the first backup we made with ID ```b19544cf```, all we have to do is:

```
restic forget b19544cf
```

the snapshot will then be removed.  It's important to note at this time, no disk space has been recovered.  To do that we need to do:

```
restic prune
```

Obviously if we do this with only one snapshot then we won't have any backups at all.  Once two backups exist, the first one can be forgotten and pruned and then the latest snapshot will be quiesced so that it includes all of the files from the first backup, that were not in the second backup.  The next time we make the next backup, we'll get another snapshot with files that changed.  Obviously you can keep a lot more snapshots where required for your backup plan.

## Restoring

Since Zimbra uses sparse files, we need to use an additional parameter during the restore process.  During my testing for restoring data, I basically deleted everything within ```/opt/zimbra``` and then started the restore process.  The restore command we need to use is as follows:

```
restic restore --sparse <id> --target /
```

so, using the result from our backups previously made and using the last backup we made (second one), the command would be:

```
restic restore --sparse 0af10032 --target /
```

that way, the sparse file will restore, but only for the amount of disk space it was using, and not the full file size if we were to attempt doing this without the parameter.  It's also questionable as to whether Zimbra would actually work afterwards (I haven't tested this).

Once Zimbra has been restored, we can then start Zimbra and see if everything comes up or not.  Let's test that now from my working example above:

First we stop Zimbra:

```
root@zimbra:~/scripts# /etc/init.d/zimbra stop
Host zimbra.example.com
	Stopping zmconfigd...Done.
	Stopping license-daemon...Done.
	Stopping zimlet webapp...Done.
	Stopping zimbraAdmin webapp...Done.
	Stopping zimbra webapp...Done.
	Stopping service webapp...Done.
	Stopping stats...Done.
	Stopping mta...Done.
	Stopping onlyoffice...Done.
	Stopping spell...Done.
	Stopping snmp...Done.
	Stopping cbpolicyd...Done.
	Stopping archiving...Done.
	Stopping opendkim...Done.
	Stopping amavis...Done.
	Stopping antivirus...Done.
	Stopping antispam...Done.
	Stopping proxy...Done.
	Stopping memcached...Done.
	Stopping mailbox...Done.
	Stopping logger...Done.
	Stopping dnscache...Done.
	Stopping ldap...Done.
```

Then we cleanup the existing Zimbra directory:

```
root@zimbra:~/scripts# cd /opt/zimbra

root@zimbra:/opt/zimbra# rm -rf *

root@zimbra:/opt/zimbra# ls -lha
total 36K
drwxr-xr-x 1 root   root    252 Jan 17 10:47 .
drwxr-xr-x 1 root   root     28 Jan 16 12:20 ..
-rw-r----- 1 zimbra zimbra  165 Jan 17 10:33 .bash_history
-rw-r--r-- 1 zimbra zimbra  350 Jan  1 17:36 .bash_profile
-rw-r--r-- 1 zimbra zimbra  835 Jan  1 17:36 .bashrc
-rw-r--r-- 1 zimbra zimbra   62 Jan  1 17:36 .exrc
-rw-r--r-- 1 root   root   4.3K Jan 16 21:00 .install_history
-rw-r--r-- 1 zimbra zimbra   52 Jan  1 17:36 .ldaprc
-rw-r--r-- 1 zimbra zimbra   19 Jan  1 17:36 .platform
drwxr-x--- 1 zimbra zimbra    0 Jan 16 20:38 .saveconfig
drwx------ 1 zimbra zimbra   98 Jan 16 21:00 .ssh
-rw-r--r-- 1 zimbra zimbra    0 Jan  1 17:36 .viminfo
-rw-r----- 1 zimbra zimbra    0 Jan 16 20:41 .zmmailbox_history
-rw-r----- 1 zimbra zimbra 1.7K Jan 16 20:49 .zmprov_history

root@zimbra:/opt/zimbra# rm -rf .bash* .exrc .install_history .ldaprc .platform .s* .viminfo .zm*

root@zimbra:/opt/zimbra# ls -lha
total 0
drwxr-xr-x 1 root root  0 Jan 17 10:47 .
drwxr-xr-x 1 root root 28 Jan 16 12:20 ..
```

Now we can start the restore process:

```
root@zimbra:/opt/zimbra# cd /root/scripts/

root@zimbra:/opt/zimbra# restic snapshots
repository 5339cb72 opened (version 2, compression level auto)
ID        Time                 Host        Tags        Paths
------------------------------------------------------------------
b19544cf  2024-01-17 10:21:15  zimbra                  /opt/zimbra
0af10032  2024-01-17 10:33:54  zimbra                  /opt/zimbra
------------------------------------------------------------------
2 snapshots

root@zimbra:~/scripts# restic restore --sparse 0af10032 --target /
repository 5339cb72 opened (version 2, compression level auto)
[0:00] 66.67%  2 / 3 index files loaded
restoring <Snapshot 0af10032 of [/opt/zimbra] at 2024-01-17 10:33:54.220787367 +0100 CET by root@zimbra> to /
Summary: Restored 23315 files/dirs (82.887 GiB) in 0:40
```

Now, let's start Zimbra:

```
root@zimbra:~/scripts# /etc/init.d/zimbra start
Host zimbra.example.com
	Starting ldap...Done.
	Starting zmconfigd...Done.
	Starting dnscache...Done.
	Starting logger...Done.
	Starting mailbox...Done.
	Starting memcached...Done.
	Starting proxy...Done.
	Starting amavis...Done.
	Starting antispam...Done.
	Starting antivirus...Done.
	Starting opendkim...Done.
	Starting snmp...Done.
	Starting spell...Done.
	Starting mta...Done.
	Starting stats...Done.
	Starting service webapp...Done.
	Starting zimbra webapp...Done.
	Starting zimbraAdmin webapp...Done.
	Starting zimlet webapp...Done.
```

Zimbra will take a little while to start, which is normal before all the services show as running.  So let's check that:

```
root@zimbra:~/scripts# /etc/init.d/zimbra status
Host zimbra.example.com
	amavis                  Running
	antispam                Running
	antivirus               Running
	dnscache                Running
	ldap                    Running
	logger                  Running
	mailbox                 Running
	memcached               Running
	mta                     Running
	opendkim                Running
	proxy                   Running
	service webapp          Running
	snmp                    Running
	spell                   Running
	stats                   Running
	zimbra webapp           Running
	zimbraAdmin webapp      Running
	zimlet webapp           Running
	zmconfigd               Running
```

So now we can verify by logging into webmail as well as webadmin if we have a working server - which I have done and all is working fine.

## Summary

So, as you can see it's possible to use restic to make online backups of Zimbra, and the system is actually recoverable.  As you can see, no problems with LDAP which can happen when making tar or rsync backups without using the appropriate sparse parameters.  Also, if the backup has been made to the local disk, you can then also combine this with using ```rsnapshot``` and then send your backups to a remote server or even ```rclone``` to send to various cloud provider-based storage.

As with all backups, your mileage may vary, so it's always important to do regular testing to make sure that your backup and recovery procedures do actually work.

**Comment by pixu2019** [comment]
Thanks. very informative article but pls add link to restic and display backup sh files with its name also zimbra version. I'm a little confused.
p.s: i google and there is "https://github.com/restic/restic" but not sure this is what you use.

**Comment by ianw1974** [comment]
The main page for restic: https://restic.net/
Their github is the same as you posted, also linked on their main page.

The content of the script I showed at the beginning of the post, and later used it in the tests as shown ```root@zimbra:~/scripts# ./restic-backup-zimbra-online.sh ```

I have modified the article to show the name before the script contents.  The name of the script can be pretty much anything you want it to be.

**Reply by pixu2019** [reply]
Thanks! Nice weekend!

**Comment by pixu2019** [comment]
do you have guide to migrate zimbra OSE 8/centos 6 to zimbra 10 FOSS/rocky linux( or ubuntu) using restic or any scripts? Thanks

**Comment by ianw1974** [comment]
Hi,

No I don't, server migration guides can be found in Zimbra's wiki, so the majority of stuff applies there.  See:

https://wiki.zimbra.com/wiki/How_to_move_ZCS_to_another_server

or:

https://wiki.zimbra.com/wiki/Zimbra_to_Zimbra_Migration

The first link explains migration using tools like rsync, which you can substitute for the commands above relating to restic.  So based on the wiki guide you do something like:

1. Restore zimbra using restic instead of rsync to copy between servers.
2. Run install.sh -s to get Zimbra installed but without running configuration, etc.

Then follow the Zimbra guide for remaining steps.  What is important is, that the version of Zimbra installed on the new server is the same version as what you have on CentOS.  Then once this migration has finished and you have 8.8.15 on the new server (assuming CentOS 6 is running 8.8.15), then you can think about upgrading to 10.0.6.  I suggest testing it on a copy of your CentOS 6 server.

The second link does a migration by copying/moving accounts from the old server to the new one.  So there is another option in case the first one doesn't work for you.  It does mean of course, configuring the new server for any additional custom config that you applied to the old server though.

I don't offer Zimbra support, since the Zimbra forums are the best place to search for assistance with migrating etc.  Whether you do it with restic or rsync is up to you.  The post here about restic was something I found to be useful that isn't mentioned anywhere on Zimbra forums, perhaps because nobody has any experience of it, perhaps they prefer other backup tools, or just simply not thought of trying restic as an alternative backup solution.  Either way, it's easy enough to adapt Zimbra's documentation as given in my example above for restoring with restic instead of using rsync.

**Reply by pixu2019** [reply]
Thanks you!

**Reply by ianw1974** [reply]
Just remember, use the latest version of restic which has sparse file support:

```
root@zimbra:/home/ian# restic version
restic 0.16.3 compiled with go1.21.6 on linux/amd64
```

this is the one I have used.  Hopefully that version runs on CentOS 6.
